### ISSUE TITLE
Proposal (& implementation): LINQ APIs for C# 8.0 index and range syntax

### ISSUE THREAD

**Issue by Dixin** [dotnet/corefx#35552]
C# 8.0 introduces index and range features. It is natural and convenient to support index and range in LINQ. I looked at the API review notes https://github.com/dotnet/apireviews, didn't find such APIs, so propose it here.

# Rationale and usage

Currently, the index and range work with array. The goals to work with LINQ are:
- Use index to locate an element in sequence, use range to slice sequence. The usage is the same as index/range for array, but with deferred execution for slice with range.
- Use range to start fluent LINQ query.

Currently LINQ has `ElementAt(int index)` and `ElementOrDefault(int index)` query operator. It would be natural to have a overload for `System.Index`: `ElementAt(Index index)` and `ElementOrDefault(Index index)`, and a new method `ElementsIn(Range range)`, so that LINQ can seamlessly work with C# 8.0:
```cs
Index index = ...;
var element1 = source1.ElementAt(index);
var element2 = source2.ElementAtOrDefault(^5);
Range range = ...;
var slice1 = source3.ElementsIn(range);
var slice2 = source4.ElementsIn(2..^2)
var slice2 = source5.ElementsIn(^10..);
```

The following `Range` overload and `AsEnumerable` overload for `System.Range` convert it to a sequence, so that LINQ query can be started fluently from c# range:
```cs
var query1 = Enumerable.Range(10..).Select(...);
Range range = ...;
var query1 = range.AsEnumerable().Select(...);
var query2 = (10..20).AsEnumerable().Where(...);
```

# Proposed APIs

For LINQ to Objects:
```cs
namespace System.Linq
{
    public static partial class Queryable
    {
        public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, Index index) { throw null; }

        public static TSource ElementAtOrDefault<TSource>(this IEnumerable<TSource> source, Index index) { throw null; }

        public static IEnumerable<TSource> ElementsIn<TSource>(this IEnumerable<TSource> source, Range range) { throw null; }

        public static IEnumerable<TSource> Range<TSource>(Range range) { throw null; }

        public static IEnumerable<TSource> AsEnumerable<TSource>(this Range source) { throw null; }
}
```
For remote LINQ:
```cs
namespace System.Linq
{
    public static partial class Queryable
    {
        public static TSource ElementAt<TSource>(this IQueryable<TSource> source, Index index) { throw null; }

        public static TSource ElementAtOrDefault<TSource>(this IQueryable<TSource> source, Index index) { throw null; }

        public static IQueryable<TSource> ElementsIn<TSource>(this IQueryable<TSource> source, Range range) { throw null; }
}
```

# Implementation details (and pull request)

The [API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) says PR should not be submitted before the API proposal is approved. So currently I implemented these APIs separately:
- `Enumerable`: [`ElementAt(Index)`, `ElementOrDefaultAt(Index)`](https://github.com/Dixin/CodeSnippets/blob/master/Linq.Range/Linq.Range/ElementAt.cs)
- `Enumerable`: [`ElementsIn(Range)`](https://github.com/Dixin/CodeSnippets/blob/master/Linq.Range/Linq.Range/ElementsIn.cs)
- `Enumerable`: [`Range(Range)`](https://github.com/Dixin/CodeSnippets/blob/master/Linq.Range/Linq.Range/Range.cs) and [`AsEnumerable(Range)`](https://github.com/Dixin/CodeSnippets/blob/master/Linq.Range/Linq.Range/AsEnumerable.cs). They are the same convertion.
- `Queryable`: [`ElementAt(Index)`, `ElementOrDefaultAt(Index)`, `ElementsIn(Range)`](https://github.com/Dixin/CodeSnippets/blob/master/Linq.Range/Linq.Range/QueryableExtensions.cs).

These APIs' implementation is self-contained so that the code can be just copied to use. If this proposal is doable, I can submit a PR quickly.

The implementation of `ElementAt(Index)`, `ElementOrDefault(Index)` and `ElementsIn(Range)` for `IQueryable<T>` is straightforward. They just create an expression tree. So are these methods for `IEnumerable<T>`. I just followed the behavior and exceptions of array with range (See unit tests [here](https://github.com/Dixin/CodeSnippets/blob/master/Linq.Range/Linq.Range.Tests/ElementsInTests.cs)).

# Open questions

For `Range(Range)` and `AsEnumerable(Range)`, the question is: what does `Range`'s start `Index` and end `Index` mean when the index is from the end? For example, `10..20` can be easily converted to a sequence of 10, 11,12, ... 19, but how about `^20...^10`? In my current implementation, regarding `Index`'s value can be from `0` to `int.MaxValue`, I assume a virtual "full range" `0..2147483648`, and any `Range` instance is a slice of that "full range". So:

- Ranges `..` and `0..` and `..^0` and `0..^0` are converted to "full sequence" 0, 1, .. 2147483647
- Range `100..^47` is converted to sequence 100, 101, .. 2147483600
- Range `^48..^40` is converted to sequence 2147483600, 2147483601 .. 2147483607
- Range `10..10` is converted to empty sequence

etc. See unit tests of `AsEnumerable(Range)` [here](https://github.com/Dixin/CodeSnippets/blob/master/Linq.Range/Linq.Range.Tests/AsEnumerableTests.cs).

**Comment by bartdesmet** [comment]
This looks interesting to me and has been on my list of things to look into for Ix and async LINQ for `IAsyncEnumerable<T>` over at https://github.com/dotnet/reactive/pull/423. Adding @terrajobst for visibility, given he owns `System.Index` and `System.Range` and is also aware of the work in Ix.

From a technical point of view, here are my 0.02 cents:

# `ElementAt` and `ElementAtOrDefault`

These do make sense and have little ambiguity. Indexing from the end is also not a concern, given `TakeLast` was also added to LINQ surface before. It'd be nice to define some algebraic identities, e.g. converting `ElementAt[OrDefault](i)` to `Skip(n).First[OrDefault]()` if indexing from the start or `SkipLast(n).Last[OrDefault]()` if indexing from the end, at least for purposes of reviewing a design (and to write tests to assert the identity, though likely not for an implementation when it comes to efficiency).

# `ElementsIn`

This one is a bit more fuzzy. At first glance, it feels like a `Slice` method and obeys, as a first approximation, to the identity `xs.Slice(r).SequenceEqual(xs.ToArray().Slice(r))`, except for infinite sequences where `ToArray` wouldn't terminate. In this case, there'd be exceptions thrown for range violations. Likely `ArgumentOutOfRangeException` to match existing LINQ operators, rather than `OverflowException` or something like that.

Alternatively, the identity could be defined in terms of composing `Skip[Last]` and `Take[Last]`. However, that'd bring with it other exception behavior (namely, none), because `Skip` and `Take` are happy to yield empty sequences.

So the question would be whether to lean towards array semantics or LINQ semantics, and an argument could be made for both sides (though, also based on the `Range` discussion below, I'd be leaning towards the former).

# `Range` and `AsEnumerable`

A first question is whether `Range` implementing `IEnumerable<int>` would make sense. Most likely not, because "from end" values. Would these be relative to `int.MaxValue` (as in the open question raised here), which yields integers that are not relative to some list-like thing being indexed into (which is the primary use of `Range`)? Given this observation, I'm not the biggest fan of an `AsEnumerable` extension method on `Range` because it could/would give the false illusion of "if Range were to implement IEnumerable<int>, it'd do it like this".

As for `Enumerable.Range`, that could make sense given it would be clearer than `start, count` pairs where one often wonders if the second argument to `Range` is an index or a count (`..` eliminates this confusion). I actually always use named parameters when calling `Range` to clarify the meaning of the two integer arguments.

Supporting "from end" is a different matter though. On the one hand, given the `int` return type, one could "suspect" a range `0..int.MaxValue`, but a more formally inclined person could argue about the signed nature of `int` (`int.MinValue..int.MaxValue`?). I'd personally not be offended by `Range` throwing `Argument[OutOfRange]Exception` to indicate there's no "from end" support, and just have it work for cases that would already be expressible using current the `Enumerable.Range` method, i.e. relative to `0`. Open-ended ranges are still a question then, either picking `int.MaxValue` or throwing.

One could argue that `Range(r).Select(i => xs.ElementAt(i)).SequenceEqual(xs.ElementsIn(r))` should be an identity (modulo side-effects of repeated enumeration due to `ElementAt`), thus values relative to `int.MaxValue` would not make sense to be treated as indexes passed to `ElementAt` (i.e. the view where an enumeration of a range yields valid indexes, which it can't do for "from end" indexes if the length of the source is unknown, as is the case for arbitrary `IEnumerable<T>` objects). That'd also cement `ElementsIn` (or maybe `Slice` if that's a better name) behavior of throwing exceptions on range violations (rather than yielding an empty sequence).

# `Queryable` methods

Symmetry makes sense to me; there's been a precedent of extending the surface with `TakeLast` and `SkipLast`. However, two possible concerns could be:

* Query providers may break in a variety of ways because they see new overloads to `ElementAt[OrDefault]`. E.g. they could use reflection to get `MethodInfo` for these, under the assumption there's only a single overload (e.g. short-sighted use of `Single` or `[0]`). Or, they could match simply on name and try to interpret the second argument as `Int32` and fail there. Note this is different from `TakeLast` and `SkipLast` which were brand new names.
* Query providers could freak out seeing `Index` or `Range` objects appear in an expression tree. I'm also not sure on the plan of record for using `..` in lambda expressions converted to expression trees; either these would become `MethodCallExpression` nodes to factory methods (as is the case for `$"..."` interpolated strings using `string.Format`), or get rejected by the C# compiler (as is the case for e.g. `?.`).

One could argue different ways here:

- `.NET Core` supports side-by-side, so this type of compat is less of a concern today. We could also make sure that `LINQ to SQL` and Entity Framework are not affected.
- We can't run the risk, and
  - asymmetry is tolerable (which makes me cringe);
  - we can put it in another library (thus on another `EnumerableExtensions` method, reducing the risk of query providers blowing up; though if they just check method names, they'd still be in trouble).

For the very last (personally not preferred, to be clear) option, note that the query provider supporting `AsQueryable` on an `IEnumerable<T>` is happy to find an `IEnumerable<T>`-based method on the same declaring type for purposes of rebinding `IQueryable<T>`-based methods to their "local" equivalent (see https://github.com/dotnet/corefx/blob/master/src/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs#L49), so the symmetry can be kept in an extension library (such as Ix).

One data point for this decision could be the proposed addition of a `Zip` method with a `ValueTuple` return type, see https://github.com/dotnet/corefx/issues/16011. This is analogous in two ways: reuse of the same method name, and occurrence of a new type in expression trees (though not in an operator argument position, but in the return type, so typically treated differently by query providers).

**Comment by Clockwork-Muse** [comment]
In light of symmetry, don't forget PLINQ (where operators must be implemented by the library, and not with add-ons).

**Comment by Dixin** [comment]
@bartdesmet Thank you for your insights. I hope these LINQ operators can be provided in corefx. Since index/range are language level features, it is nice to have them work with not only array, but also any type that implements `IEnumerable<T>`.

**Comment by Dixin** [comment]
@Clockwork-Muse I looked at PLINQ before I post this proposal. `ParallelEnumerable` is not symmetric with `Enumerable`/`Queryable`. `Enumerable`/`Queryable` have `Append`/`Prepend`/`SkipLast`/`TakeLast` operators but `ParallelEnumerable` does not have them.

**Comment by Clockwork-Muse** [comment]
@Dixin - 
That's fine, if you've already reviewed it and decided not to include it.

**Comment by joshfree** [comment]
@tarekgh

**Comment by Dixin** [comment]
@joshfree @tarekgh Thank you for labeling this issue. Since it is now in the Future Milestone ([issues not triaged to a planned release. Work in this milestone may happen, or may not](https://github.com/dotnet/corefx/milestone/5)). Does it mean there is not much chance of actions for the next release? Please suggest what should I do next.

**Comment by tarekgh** [comment]
> Does it mean there is not much chance of actions for the next release? Please suggest what should I do next.

We'll look at the details to ensure the proposed APIs is reasonable and then we'll have the issue reviewed by the design committee. expect will get traction on this when we ship 3.0. if you didn't see any activities by then, please ping us. Usually we'll get into that in right time. Thanks for the proposal.

**Comment by Dixin** [comment]
Thank you @tarekgh. I will wait by then.
